sdk/src/main/scala/com/horizen/storage/SidechainStateStorage.scala

```diff
@@ -103,13 +120,30 @@ class SidechainStateStorage (storage : Storage, sidechainBoxesCompanion: Sidecha
       new ByteArrayWrapper(WithdrawalEpochInfoSerializer.toBytes(withdrawalEpochInfo))))
 
     if (withdrawalRequestAppendList.nonEmpty) {
-      val withdrawalRequestList = getWithdrawalRequests(withdrawalEpochInfo.epoch)
-
-      withdrawalRequestList.addAll(withdrawalRequestAppendList.asJavaCollection)
+      val withdrawalRequestList = getWithdrawalRequests(withdrawalEpochInfo.epoch) ++ withdrawalRequestAppendList
 
       updateList.add(new JPair(getWithdrawalRequestsKey(withdrawalEpochInfo.epoch),
-        new ByteArrayWrapper(withdrawalRequestSerializer.toBytes(withdrawalRequestList))))
+        new ByteArrayWrapper(withdrawalRequestSerializer.toBytes(withdrawalRequestList.asJava))))
+
+    }
 
+    for (btc <- mainchainBTCList.filter(_.epochNumber == withdrawalEpochInfo.epoch - 1)) {
```

Comment:
Certificate validation should be done in `SidechainState.validate(mod)` method. At the moment there is no certificate validation at all there.

Certificate validation should be done in both SidechainHistory and SidechainState:
During block appending to History additional checks in `WithdrawalEpochValidator` should be done (or even new `CertificateValidator` can be introduce):
1. It should be 0 or 1 Certificate inside included MCBlockReferences
2. That certificate should be for WithdrawalEpoch N-1, where N is withdrawalEpoch number of the block applied.
3. Verify certificate `sidechainId` equals to ours.
4. Verify certificate `endEpochBlockHash` should be the last mc block ref id of the last scblock in the previous withdrawalEpoch for the applied block chain.

So considering the checks in the History in `SidechainState.validate(mod)` you should check:
1. Certificate withdrawals list equals to the withdrawals list stored in the State for that epoch (like you did below)
2. Possibly verify that it's the first certificate applied  for current branch for the withdrawal epoch.
NOTE: If we need to check (2), you can pass some flag of certificate presence to State.applyChanges, and store it for withdrawal epoch N. And get that flag for check. So no need of concepts of  `UnprocessedWithdrawalRequests` and `getWithdrawalRequests`, no need of data duplication. Moreover, this flag can be used by Submitter later to decide to act or not.

Reply:
Note: Non-valid certificate leads to non-valid block